edifice/engine.py

```diff
@@ -621,6 +624,25 @@ def _render(self, component: Element, render_context: _RenderContext):
         component._hook_state_index = 0
         component._hook_effect_index = 0
         component._hook_async_index = 0
+
+        # Before the render, reduce the _hook_state updaters.
+        # We can't do this after the render, because there may have been state
+        # updates from event handlers.
+        for hooks in self._hook_state.values():
+            for hook in hooks:
+                state0 = hook.state
+                try:
+                    for updater in hook.updaters:
+                        if callable(updater):
```

Comment:
Oh no I just realized that this won't work because the user might be storing a callable in the state.

Reply:
I think that's fine. We could document that it needs to wrapped in a data structure like a tuple if you need to store a function.

Reply:
Good idea.